crates/ruff_python_ast/src/source_code/line_index.rs

```diff
@@ -236,49 +236,71 @@ impl IndexKind {
 /// Type-safe wrapper for a value whose logical range starts at `1`, for
 /// instance the line or column numbers in a file
 ///
-/// Internally this is represented as a [`NonZeroUsize`], this enables some
+/// Internally this is represented as a [`NonZeroU32`], this enables some
 /// memory optimizations
+///
+/// Conversion between `usize` and `OneIndexed` always specify
+/// zero-indexed or one-indexed. The internal representation is always
+/// one-indexed.
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-pub struct OneIndexed(NonZeroUsize);
+pub struct OneIndexed(NonZeroU32);
+
+#[allow(clippy::cast_possible_truncation)] // manually checked
+const fn index_to_u32(value: usize) -> u32 {
+    assert!(
+        value <= u32::MAX as usize,
+        "Value is too large to be represented as a OneIndexed"
+    );
+    value as u32
+}
```

Comment:
I got the second reason, but have a question about the first reason.

What happens if source code is bigger than 4GB? Because TextSize is backed by u32, I thought source code bigger than 4GB will not be located.
Does ruff use different mechanism to solve large files?
Whether it is or not, the exact 4GB file seem to be handled in same way as 4GB+ files.

----

Not strictly related to this topic, but some platforms even take 64bit file size when the underlying architecture is 32bit. Source code length can be not limited to u32. (Though I hope nobody tries to make 4GB source code in single file)

Reply:
> What happens if source code is bigger than 4GB? Because TextSize is backed by u32, I thought source code bigger than 4GB will not be located.

Ruff will panic for files larger than 4GB. We should probably add a nice diagnostic right before parsing the file to gracefully handle this case.

The main point here is that `TextSize` can handle files up to `u32::MAX`, but `OneIndexed` only up to `u32::MAX - 1`.

Reply:
That's exactly what I meant to. If ruff reject 4GB+ files, ruff can reject the exact 4GB file too when reading file if length  = u32::MAX. If ruff is going to add a nice diagnostic, it can add the same thing by checking length == u32::MAX when reading file length.
It will simplify exception handling just by excluding one very rare edge case.